File: boj/python/two_pointer/boj14719.py
import sys

input = sys.stdin.readline
h, w = map(int, input().rstrip().split(" "))
heights = list(map(int, input().rstrip().split(" ")))

left, left_height_max = 0, heights[0]
right, right_height_max = len(heights) - 1, heights[-1]
answer = 0

while left < right:
    left_height_max = max(left_height_max, heights[left])
    right_height_max = max(right_height_max, heights[right])

    if left_height_max < right_height_max:
        answer += left_height_max - heights[left]
        left += 1
    else:
        answer += right_height_max - heights[right]
        right -= 1

print(answer)

File: boj/java/src/two_pointer/Boj14719.java
package two_pointer;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;

public class Boj14719 {
	public int solve(int width, int[] heights) {
		int left = 0, right = width - 1;
		int answer = 0;

		int leftMax = heights[left];
		int rightMax = heights[right];

		while (left < right) {
			leftMax = Math.max(leftMax, heights[left]);
			rightMax = Math.max(rightMax, heights[right]);

			if (leftMax < rightMax)
				answer += leftMax - heights[left++];
			else
				answer += rightMax - heights[right--];
		}
		return answer;
	}

	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		String[] sizes = br.readLine().split(" ");

		int w = Integer.parseInt(sizes[1]);
		int[] heights = Arrays.stream(br.readLine().split(" "))
				.mapToInt(Integer::parseInt)
				.toArray();

		System.out.println(new Boj14719().solve(w, heights));
	}
}